File: test/ModularPipelines.UnitTests/Execution/DisposableModuleTests.cs
using ModularPipelines.Context;
using ModularPipelines.Modules;
using ModularPipelines.TestHelpers;

namespace ModularPipelines.UnitTests.Execution;

public class DisposableModuleTests
{
    [Test]
    public async Task SuccessfullyDisposed()
    {
        var pipelineSummary = await TestPipelineHostBuilder.Create()
            .AddModule<DisposableModule>()
            .ExecutePipelineAsync();
        await Assert.That(pipelineSummary.Modules.OfType<DisposableModule>().Single().IsDisposed).IsTrue();
    }

    public class DisposableModule : Module<bool>, IDisposable
    {
        public bool IsDisposed { get; private set; }

        /// <inheritdoc/>
        public override async Task<bool> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            // Reduced delay from 100ms to 1ms for faster test execution
            await Task.Delay(1, cancellationToken);
            return true;
        }

        /// <inheritdoc/>
        public void Dispose()
        {
            IsDisposed = true;
            GC.SuppressFinalize(this);
        }
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Execution/DisposableModuleTests.fs
namespace ModularPipelines.UnitTests.FSharp.Execution

open System
open System.Threading
open System.Threading.Tasks
open System.Linq
open ModularPipelines.Context
open ModularPipelines.Extensions
open ModularPipelines.Modules
open ModularPipelines.TestHelpers
open TUnit.Assertions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core
open TUnit.Assertions.Extensions

type private DisposableModule() =
    inherit Module<bool>()

    member val IsDisposed = false with get, set

    override _.ExecuteAsync(_: IModuleContext, cancellationToken: CancellationToken) =
        task {
            do! Task.Delay(1, cancellationToken)
            return true
        }

    interface IDisposable with
        member this.Dispose() =
            this.IsDisposed <- true
            GC.SuppressFinalize(this)

type DisposableModuleTests() =
    [<Test>]
    member _.SuccessfullyDisposed() = async {
        let! pipelineSummary =
            TestPipelineHostBuilder.Create().AddModule<DisposableModule>().ExecutePipelineAsync()
            |> Async.AwaitTask

        let isDisposed = pipelineSummary.Modules.OfType<DisposableModule>().Single().IsDisposed
        do! check(Assert.That(isDisposed).IsTrue())
    }
